In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV

import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.wrappers.scikit_learn import KerasClassifier

In [2]:
train = pd.read_csv("../Features/data_train.csv")
test = pd.read_csv("../Features/data_test.csv")
test_kaggle = pd.read_csv("../Features/data_test_kaggle.csv")

In [3]:
# x son los features, y es el target
X_train, Y_train = train.iloc[:, :-1].values, train.iloc[:, -1].values
X_test, Y_test = test.iloc[:, :-1].values, test.iloc[:, -1].values

In [4]:
X_train.shape

(13505, 61)

In [5]:
Y_test

array([1., 1., 1., ..., 1., 1., 1.])

In [6]:
norm = MinMaxScaler().fit(X_train)
X_train_norm = norm.transform(X_train)
X_test_norm = norm.transform(X_test)
test_kaggle_norm = norm.transform(test_kaggle)

In [7]:
X_train_norm

array([[3.28609056e-01, 6.73177963e-01, 3.04461942e-01, ...,
        8.56457691e-04, 9.49848024e-04, 1.00000000e+00],
       [5.78201430e-01, 7.94089811e-01, 3.36734694e-01, ...,
        2.72051267e-03, 2.26287323e-03, 1.00000000e+00],
       [6.61440696e-01, 3.79707913e-01, 5.65699659e-01, ...,
        3.02279185e-03, 4.19050599e-03, 3.77793172e-01],
       ...,
       [4.38145407e-02, 7.94089811e-01, 4.40298507e-01, ...,
        3.59712230e-03, 5.98404255e-03, 1.00000000e+00],
       [4.35636941e-02, 7.94089811e-01, 4.40298507e-01, ...,
        3.59712230e-03, 5.98404255e-03, 1.00000000e+00],
       [3.34712990e-01, 0.00000000e+00, 3.17460317e-01, ...,
        2.76701716e-03, 4.60310966e-03, 1.00000000e+00]])

In [8]:
def create_model_1(optimizer='Adamax', dropout=0.2, init='normal', n_nodes=256):
    model = Sequential()
    model.add(Dense(n_nodes, input_shape=(X_train.shape[1],), activation='relu', kernel_initializer=init))
    model.add(Dropout(dropout), )
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

In [9]:
def create_model_2(optimizer='Adamax', dropout=0.2, init='normal', n_nodes1=256, n_nodes2=256):
    model = Sequential()
    model.add(Dense(n_nodes1, input_shape=(X_train.shape[1],), activation='relu', kernel_initializer=init))
    model.add(Dropout(dropout), )
    model.add(Dense(n_nodes2, activation='relu'))
    model.add(Dropout(dropout), )
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

In [150]:
model = KerasClassifier(build_fn=create_model)

In [151]:
param_grid = {
    'epochs': [100, 1000],
    'batch_size': [1000, 5000, 10000],
    'n_nodes1': [32, 256, 512],
    'n_nodes2': [32, 256, 512],
    'dropout': [0.5, 0.4, 0.3, 0.2, 0.1, 0],
}

In [152]:
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-1, cv=4)
grid_result= grid.fit(X_train_norm, Y_train)

C:\Users\Santi\anaconda3\lib\site-packages\joblib\externals\loky\process_executor.py:688: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


Epoch 1/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.7051 - accuracy: 0.4820
Epoch 2/1000
2/2 [==============================] - 0s 40ms/step - loss: 0.6590 - accuracy: 0.5465
Epoch 3/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.6396 - accuracy: 0.5596
Epoch 4/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.6154 - accuracy: 0.6623
Epoch 5/1000
2/2 [==============================] - 0s 40ms/step - loss: 0.5987 - accuracy: 0.8105
Epoch 6/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.5817 - accuracy: 0.8307
Epoch 7/1000
2/2 [==============================] - 0s 41ms/step - loss: 0.5608 - accuracy: 0.8203
Epoch 8/1000
2/2 [==============================] - 0s 39ms/step - loss: 0.5442 - accuracy: 0.7922
Epoch 9/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.5288 - accuracy: 0.7964
Epoch 10/1000
2/2 [==============================] - 0s 39ms/step - loss: 0.5128 - accuracy: 0.8245
Epoch 11/

2/2 [==============================] - 0s 39ms/step - loss: 0.2782 - accuracy: 0.8863
Epoch 84/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2776 - accuracy: 0.8860
Epoch 85/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2770 - accuracy: 0.8855
Epoch 86/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2764 - accuracy: 0.8869
Epoch 87/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2758 - accuracy: 0.8871
Epoch 88/1000
2/2 [==============================] - 0s 35ms/step - loss: 0.2758 - accuracy: 0.8860
Epoch 89/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2746 - accuracy: 0.8869
Epoch 90/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2753 - accuracy: 0.8883
Epoch 91/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2738 - accuracy: 0.8873
Epoch 92/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2738 - accuracy: 0.8859
Epoch 93/1000


2/2 [==============================] - 0s 37ms/step - loss: 0.2443 - accuracy: 0.8980
Epoch 246/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2444 - accuracy: 0.8979
Epoch 247/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2440 - accuracy: 0.8976
Epoch 248/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2439 - accuracy: 0.8975
Epoch 249/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2440 - accuracy: 0.8980
Epoch 250/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2437 - accuracy: 0.8979
Epoch 251/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2439 - accuracy: 0.8975
Epoch 252/1000
2/2 [==============================] - 0s 35ms/step - loss: 0.2435 - accuracy: 0.8980
Epoch 253/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2436 - accuracy: 0.8988
Epoch 254/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2434 - accuracy: 0.8981
Epoch

2/2 [==============================] - 0s 37ms/step - loss: 0.2281 - accuracy: 0.9028
Epoch 408/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2284 - accuracy: 0.9041
Epoch 409/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2280 - accuracy: 0.9032
Epoch 410/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2278 - accuracy: 0.9034
Epoch 411/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2281 - accuracy: 0.9042
Epoch 412/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2282 - accuracy: 0.9027
Epoch 413/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2278 - accuracy: 0.9037
Epoch 414/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2279 - accuracy: 0.9050
Epoch 415/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2280 - accuracy: 0.9029
Epoch 416/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2275 - accuracy: 0.9041
Epoch

2/2 [==============================] - 0s 37ms/step - loss: 0.2156 - accuracy: 0.9102
Epoch 570/1000
2/2 [==============================] - 0s 35ms/step - loss: 0.2143 - accuracy: 0.9100
Epoch 571/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2146 - accuracy: 0.9096
Epoch 572/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2144 - accuracy: 0.9110
Epoch 573/1000
2/2 [==============================] - 0s 35ms/step - loss: 0.2140 - accuracy: 0.9106
Epoch 574/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2144 - accuracy: 0.9098
Epoch 575/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2143 - accuracy: 0.9114
Epoch 576/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2140 - accuracy: 0.9114
Epoch 577/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2141 - accuracy: 0.9104
Epoch 578/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2137 - accuracy: 0.9116
Epoch

2/2 [==============================] - 0s 38ms/step - loss: 0.2019 - accuracy: 0.9168
Epoch 732/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2017 - accuracy: 0.9157
Epoch 733/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2015 - accuracy: 0.9156
Epoch 734/1000
2/2 [==============================] - 0s 36ms/step - loss: 0.2013 - accuracy: 0.9161
Epoch 735/1000
2/2 [==============================] - 0s 39ms/step - loss: 0.2014 - accuracy: 0.9168
Epoch 736/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2013 - accuracy: 0.9155
Epoch 737/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.2013 - accuracy: 0.9171
Epoch 738/1000
2/2 [==============================] - 0s 39ms/step - loss: 0.2012 - accuracy: 0.9157
Epoch 739/1000
2/2 [==============================] - 0s 39ms/step - loss: 0.2012 - accuracy: 0.9165
Epoch 740/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.2011 - accuracy: 0.9177
Epoch

2/2 [==============================] - 0s 37ms/step - loss: 0.1896 - accuracy: 0.9228
Epoch 894/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.1898 - accuracy: 0.9216
Epoch 895/1000
2/2 [==============================] - 0s 38ms/step - loss: 0.1901 - accuracy: 0.9233
Epoch 896/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.1894 - accuracy: 0.9228
Epoch 897/1000
2/2 [==============================] - 0s 37ms/step - loss: 0.1894 - accuracy: 0.9238
Epoch 898/1000
2/2 [==============================] - 0s 41ms/step - loss: 0.1896 - accuracy: 0.9214
Epoch 899/1000
2/2 [==============================] - 0s 40ms/step - loss: 0.1897 - accuracy: 0.9240
Epoch 900/1000
2/2 [==============================] - 0s 40ms/step - loss: 0.1897 - accuracy: 0.9209
Epoch 901/1000
2/2 [==============================] - 0s 40ms/step - loss: 0.1894 - accuracy: 0.9238
Epoch 902/1000
2/2 [==============================] - 0s 42ms/step - loss: 0.1891 - accuracy: 0.9226
Epoch

In [153]:
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))


Best: 0.878782 using {'batch_size': 10000, 'dropout': 0, 'epochs': 1000, 'n_nodes1': 512, 'n_nodes2': 32}
0.821627 (0.045456) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 32, 'n_nodes2': 32}
0.857829 (0.012040) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 32, 'n_nodes2': 256}
0.870933 (0.030151) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 32, 'n_nodes2': 512}
0.864345 (0.012894) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 256, 'n_nodes2': 32}
0.871599 (0.033747) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 256, 'n_nodes2': 256}
0.869156 (0.032808) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 256, 'n_nodes2': 512}
0.869602 (0.017373) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 512, 'n_nodes2': 32}
0.867009 (0.025084) with: {'batch_size': 1000, 'dropout': 0.5, 'epochs': 100, 'n_nodes1': 512, 'n_nodes2': 256}
0.87

In [154]:
test_accuracy = grid.score(X_test_norm, Y_test)

1/1 [==============================] - 0s 2ms/step - loss: 0.6295 - accuracy: 0.7798


### Mejores hiperparametros para 1 hidden layer
Optimizer: Adamax  
Dropout: 0.2  
Init: Normal  
N_nodes: 512  
Batch Size: 1000  
Epochs: 100  
Resultado en Kaggle: 1.15

### Mejores hiperparametros para 2 hidden layers
Optimizer: Adamax  
Dropout: 0  
Init: Normal  
N_nodes1: 512  
N_nodes2: 32  
Batch size: 1000  
Epochs: 100  
Resultado en kaggle: 0.57

In [32]:
red_neuronal_1 = create_model_1('Adamax', 0.2, 'normal', 512)

In [33]:
history = red_neuronal_1.fit(X_train_norm, Y_train, epochs=100, batch_size=1000)

Epoch 1/100
14/14 [==============================] - 0s 19ms/step - loss: 0.6151 - accuracy: 0.6663
Epoch 2/100
14/14 [==============================] - 0s 23ms/step - loss: 0.5048 - accuracy: 0.8313
Epoch 3/100
14/14 [==============================] - 0s 21ms/step - loss: 0.4426 - accuracy: 0.8427
Epoch 4/100
14/14 [==============================] - 0s 18ms/step - loss: 0.4027 - accuracy: 0.8463
Epoch 5/100
14/14 [==============================] - 0s 16ms/step - loss: 0.3804 - accuracy: 0.8522
Epoch 6/100
14/14 [==============================] - 0s 16ms/step - loss: 0.3639 - accuracy: 0.8560
Epoch 7/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3549 - accuracy: 0.8576
Epoch 8/100
14/14 [==============================] - 0s 17ms/step - loss: 0.3467 - accuracy: 0.8598
Epoch 9/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3401 - accuracy: 0.8605
Epoch 10/100
14/14 [==============================] - 0s 17ms/step - loss: 0.3344 - accuracy: 0.8646

In [34]:
accuracy = red_neuronal_1.evaluate(X_test_norm, Y_test)

106/106 [==============================] - 0s 859us/step - loss: 0.5315 - accuracy: 0.7685


In [40]:
result = red_neuronal_1.predict(X_test_norm).astype('float64')

In [36]:
error = log_loss(Y_test, result)
error

0.5315495620698362

In [ ]:
test_pred = red_neuronal.predict(test_kaggle_norm)

In [175]:
oportunidades_test = test_kaggle['Opportunity_ID']
oportunidades_test = oportunidades_test.to_frame()

oportunidades_test['Target'] = test_pred

preds_test_agrupado_por_opp = oportunidades_test.groupby('Opportunity_ID').agg({'Target' : 'mean'})
preds_test_agrupado_por_opp = preds_test_agrupado_por_opp.reset_index()

In [176]:
preds_test_agrupado_por_opp = preds_test_agrupado_por_opp.astype({'Opportunity_ID': int})
preds_test_agrupado_por_opp

Opportunity_ID    Target
0              10689  0.564624
1              10690  0.521889
2              10691  0.269828
3              10692  0.319681
4              10693  0.513815
...              ...       ...
1562           12364  0.582690
1563           12365  0.167192
1564           12366  0.126300
1565           12367  0.300492
1566           12368  0.567527

[1567 rows x 2 columns]

In [177]:
preds_test_agrupado_por_opp.to_csv("pred_red_neuronal1.csv", index = False)

### Red Neuronal con 2 hidden layers

In [16]:
red_neuronal_2 = create_model_2('Adamax', 0, 'normal', 512, 32)

In [17]:
history = red_neuronal_2.fit(X_train_norm, Y_train, epochs=100, batch_size=1000)

Epoch 1/100
14/14 [==============================] - 0s 19ms/step - loss: 0.5600 - accuracy: 0.7582
Epoch 2/100
14/14 [==============================] - 0s 18ms/step - loss: 0.4127 - accuracy: 0.8443
Epoch 3/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3588 - accuracy: 0.8552
Epoch 4/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3343 - accuracy: 0.8612
Epoch 5/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3195 - accuracy: 0.8691
Epoch 6/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3097 - accuracy: 0.8729
Epoch 7/100
14/14 [==============================] - 0s 18ms/step - loss: 0.3031 - accuracy: 0.8753
Epoch 8/100
14/14 [==============================] - 0s 18ms/step - loss: 0.2957 - accuracy: 0.8782
Epoch 9/100
14/14 [==============================] - 0s 18ms/step - loss: 0.2903 - accuracy: 0.8795
Epoch 10/100
14/14 [==============================] - 0s 18ms/step - loss: 0.2866 - accuracy: 0.8809

In [18]:
accuracy = red_neuronal_2.evaluate(X_test_norm, Y_test)

106/106 [==============================] - 0s 859us/step - loss: 0.5608 - accuracy: 0.7803


In [19]:
result = red_neuronal_2.predict(X_test_norm).astype('float64')

In [20]:
error = log_loss(Y_test, result)
error

0.5607905576339088

### Probamos usando hyperopt

In [37]:
# Otro optimizador:
# fmin: class which will host optimization process
# tpe: the optimizer to be used (Tree of Parzen Estimator)
# hp: for defining the search space
from hyperopt import STATUS_OK
from hyperopt import fmin, tpe, hp
from hyperopt.pyll.base import scope 

def objective(x):
    model = create_model_2(optimizer='Adamax', dropout=x['dropout'], init='normal',\
                         n_nodes1=x['n_nodes1'], n_nodes2=x['n_nodes2'])
    history = model.fit(X_train_norm, Y_train, epochs=100, batch_size=x['batch_size'], verbose=0)
    result = model.predict(X_test_norm)
    logloss = log_loss(Y_test, result)
    # print(logloss)
    return {
    'loss': logloss,
    'status': STATUS_OK
  }

# Search space definition
# space = hp.uniform('x', -2, 2)
space = {'dropout' : scope.float(hp.uniform('dropout',0,0.5)),
         'n_nodes1' : scope.int(hp.uniform('n_nodes1',30,1000)),
         'n_nodes2' : scope.int(hp.uniform('n_nodes2',30,1000)),
         'batch_size' : scope.int(hp.uniform('batch_size',100, 2000))
        }

# Init the optimizer
# fn: function to be optimized
# space: search space
# algo: optimizer algorithm
# max_evals: number of iterations
best = fmin(
    fn=objective,  
    space=space,
    algo=tpe.suggest,
    max_evals=500
)

print(f"Optimal value of x: {best}")

100%|█| 1000/1000 [39:43:13<00:00, 142.99s/trial, best loss: 0.4604474979747771
Optimal value of x: {'batch_size': 1497.0161707995708, 'dropout': 0.06161599566644749, 'n_nodes1': 789.9833763815055, 'n_nodes2': 180.622335180733}


### Mejores hiperparametros con hyperopt

Batch Size: 1497  
Dropout: 0.06  
n_nodes1: 790  
n_nodes2: 181  
Epochs: 100  
Optimizer: Adamax  
Init: Normal  

Optimal value of x: {'batch_size': 1497.0161707995708, 'dropout': 0.06161599566644749, 'n_nodes1': 789.9833763815055, 'n_nodes2': 180.622335180733}

In [66]:
model = create_model_2('Adamax', 0.06161599566644749, 'normal', 790, 181)

In [67]:
history = model.fit(X_train_norm, Y_train, epochs=50, batch_size=1497)

Epoch 1/50
10/10 [==============================] - 1s 62ms/step - loss: 0.5707 - accuracy: 0.7338
Epoch 2/50
10/10 [==============================] - 1s 62ms/step - loss: 0.4159 - accuracy: 0.8432
Epoch 3/50
10/10 [==============================] - 1s 60ms/step - loss: 0.3627 - accuracy: 0.8531
Epoch 4/50
10/10 [==============================] - 1s 64ms/step - loss: 0.3435 - accuracy: 0.8585
Epoch 5/50
10/10 [==============================] - 1s 61ms/step - loss: 0.3294 - accuracy: 0.8646
Epoch 6/50
10/10 [==============================] - 1s 63ms/step - loss: 0.3292 - accuracy: 0.8615
Epoch 7/50
10/10 [==============================] - 1s 62ms/step - loss: 0.3184 - accuracy: 0.8703
Epoch 8/50
10/10 [==============================] - 1s 60ms/step - loss: 0.3122 - accuracy: 0.8718
Epoch 9/50
10/10 [==============================] - 1s 60ms/step - loss: 0.3055 - accuracy: 0.8743
Epoch 10/50
10/10 [==============================] - 1s 63ms/step - loss: 0.3005 - accuracy: 0.8769
Epoch 11/

In [68]:
result = model.predict(X_test_norm).astype("float64")

In [69]:
error = log_loss(Y_test, result)
error

0.4924576111812706

In [60]:
test_pred = model.predict(test_kaggle_norm)

In [61]:
oportunidades_test = test_kaggle['Opportunity_ID']
oportunidades_test = oportunidades_test.to_frame()

oportunidades_test['Target'] = test_pred

preds_test_agrupado_por_opp = oportunidades_test.groupby('Opportunity_ID').agg({'Target' : 'mean'})
preds_test_agrupado_por_opp = preds_test_agrupado_por_opp.reset_index()

In [62]:
preds_test_agrupado_por_opp = preds_test_agrupado_por_opp.astype({'Opportunity_ID': int})
preds_test_agrupado_por_opp

Opportunity_ID    Target
0              10689  0.711828
1              10690  0.591198
2              10691  0.730097
3              10692  0.501325
4              10693  0.481910
...              ...       ...
1562           12364  0.589720
1563           12365  0.725122
1564           12366  0.665753
1565           12367  0.737231
1566           12368  0.699353

[1567 rows x 2 columns]

In [63]:
preds_test_agrupado_por_opp.to_csv("pred_red_neuronal1.csv", index = False)